lib/gc/Transforms/AnyTilableFusion.cpp

```diff
@@ -0,0 +1,717 @@
+        alreadyTiledOps.insert(tiledOp);
+        anchorOpList.push_back(tiledOp);
+      }
+    }
+    // post-op fuse
+    for (OpResult result : anchorOp->getResults()) {
+      auto tiledOpList = postOpFuseConsumerOfOpResult(rewriter, loc, result,
+                                                      desc, alreadyTiledOps);
+      for (auto &tiledOp : tiledOpList) {
+        alreadyTiledOps.insert(tiledOp);
+        anchorOpList.push_back(tiledOp);
+      }
+    }
+  }
+}
+
+/**
+ * What is Anchor Op?
+ * 1. located in a for loop
+ * 2. it is the only one TilingInterface op in for loop
+ * 3. has extract/insert slice
+ *
+ * E.g.
+ * %1 = scf.for(){
+ *   %2 = scf.for(){
+ *       %3 = extract_slice
+ *       %4 = anchor_op(%3)
+ *       %5 = insert %4
+ *       yield %5
+ *   }
+ * }
+ *
+ * */
+static LogicalResult isAnchorOp(Operation *targetOp) {
+  // 0. check tilable
+  if (!isa<TilingInterface>(targetOp)) {
+    return failure();
+  }
+  // 1. check parentOp
+  auto forOp = targetOp->getParentOfType<LoopLikeOpInterface>();
+  if (!forOp) {
+    return failure();
+  }
+  // 2. check single one tiling interface in loop body
+  auto walkResult = forOp->walk([&targetOp](TilingInterface op) {
+    // some special op maybe already deal with in template
+    if (isa<linalg::FillOp>(op))
+      return WalkResult::skip();
+    return op != targetOp ? WalkResult::interrupt() : WalkResult::advance();
+  });
+  if (walkResult.wasInterrupted()) {
+    return failure();
+  }
+  // 3. check whether has either extract or insert slice op
+  walkResult = forOp->walk(
+      [](tensor::ExtractSliceOp) { return WalkResult::interrupt(); });
+  if (walkResult.wasInterrupted()) {
+    return success();
+  }
+  walkResult = forOp->walk(
+      [](tensor::InsertSliceOp) { return WalkResult::interrupt(); });
+  return success(walkResult.wasInterrupted());
+}
+
+static bool TilingAndFusionBasedOnAnchorOp(RewriterBase &rewriter, Location loc,
+                                           func::FuncOp f, SystemDesc desc) {
+  SmallVector<Operation *> anchorOpList;
+  // Walk through func operation.
+  f->walk([&anchorOpList](Operation *op) {
+    if (succeeded(isAnchorOp(op))) {
+      anchorOpList.push_back(op);
+    }
+  });
+
+  for (auto &anchorOp : anchorOpList) {
+    diffusion(rewriter, f->getLoc(), anchorOp, desc);
+  }
+
+  // Return whether need repartition
+  return true;
+}
+
+static void FineGrainedFusion(RewriterBase &rewriter, Location loc,
+                              func::FuncOp f, SystemDesc desc) {
+  // Target at anchor op, like matmul/conv, and try to fuse any tilable
+  // operation around it by diffusion
+  TilingAndFusionBasedOnAnchorOp(rewriter, loc, f, desc);
+}
+
+struct AnyTilableFusion : public impl::AnyTilableFusionBase<AnyTilableFusion> {
+
+public:
+  void runOnOperation() final {
+    auto &ctx = getContext();
+
+    func::FuncOp func = getOperation();
```

Comment:
Why here directly return `FuncOp`? I see this operation only return base class operation? Or may be we can use `PassWrapper<impl::AnyTilableFusionBase<AnyTilableFusion>,
          OperationPass<func::FuncOp>>` to staticly filter out func operation?